Source/Core/Core/HW/WiiSave.cpp

```diff
@@ -41,525 +48,509 @@ constexpr Md5 s_md5_blanker{{0x0E, 0x65, 0x37, 0x81, 0x99, 0xBE, 0x45, 0x17, 0xA
-  const IOS::CertECC device_certificate = m_ios.GetIOSC().GetDeviceCertificate();
-  data_file.WriteArray(&device_certificate, 1);
-  data_file.WriteArray(&ap_cert, 1);
+  IOS::HLE::IOSC& m_iosc;
+  File::IOFile m_file;
+};
 
-  m_valid = data_file.IsGood();
+StoragePointer MakeNandStorage(IOS::HLE::FS::FileSystem* fs, u64 tid)
+{
+  // fs parameter is not used yet but will be after WiiSave is migrated to the new FS interface.
+  return StoragePointer{new NandStorage{tid}};
 }
 
-bool WiiSave::getPaths(bool for_export)
+StoragePointer MakeDataBinStorage(IOS::HLE::IOSC* iosc, const std::string& path, const char* mode)
 {
-  if (m_title_id)
-  {
-    // CONFIGURED because this whole class is only used from the GUI, not directly by games.
-    m_wii_title_path = Common::GetTitleDataPath(m_title_id, Common::FROM_CONFIGURED_ROOT);
-  }
+  return StoragePointer{new DataBinStorage{iosc, path, mode}};
+}
 
-  if (for_export)
-  {
-    char game_id[5];
-    sprintf(game_id, "%c%c%c%c", (u8)(m_title_id >> 24) & 0xFF, (u8)(m_title_id >> 16) & 0xFF,
-            (u8)(m_title_id >> 8) & 0xFF, (u8)m_title_id & 0xFF);
+template <typename T>
+static bool Copy(const char* description, Storage* source, std::optional<T> (Storage::*read_fn)(),
+                 Storage* dest, bool (Storage::*write_fn)(const T&))
+{
+  const std::optional<T> data = (source->*read_fn)();
+  if (data && (dest->*write_fn)(*data))
+    return true;
+  ERROR_LOG(CORE, "WiiSave::Copy: Failed to %s %s", !data ? "read" : "write", description);
+  return false;
+}
 
-    if (!File::IsDirectory(m_wii_title_path))
-    {
-      m_valid = false;
-      ERROR_LOG(CONSOLE, "No save folder found for title %s", game_id);
-      return false;
-    }
+bool Copy(Storage* source, Storage* dest)
+{
+  return Copy("header", source, &Storage::ReadHeader, dest, &Storage::WriteHeader) &&
+         Copy("bk header", source, &Storage::ReadBkHeader, dest, &Storage::WriteBkHeader) &&
+         Copy("files", source, &Storage::ReadFiles, dest, &Storage::WriteFiles);
+}
 
-    if (!File::Exists(m_wii_title_path + "/banner.bin"))
-    {
-      m_valid = false;
-      ERROR_LOG(CONSOLE, "No banner file found for title %s", game_id);
-      return false;
-    }
-    m_encrypted_save_path += StringFromFormat("/private/wii/title/%s/data.bin", game_id);
-    File::CreateFullPath(m_encrypted_save_path);
+bool Import(const std::string& data_bin_path)
+{
+  IOS::HLE::Kernel ios;
+  const auto data_bin = MakeDataBinStorage(&ios.GetIOSC(), data_bin_path, "rb");
+  const std::optional<Header> header = data_bin->ReadHeader();
+  if (!header)
+  {
+    ERROR_LOG(CORE, "WiiSave::Import: Failed to read header");
+    return false;
   }
-  else
+  const auto nand = MakeNandStorage(ios.GetFS().get(), header->hdr.tid);
+  if (nand->SaveExists() && !AskYesNoT("Save data for this title already exists. Consider backing "
+                                       "up the current data before overwriting.\nOverwrite now?"))
```

Comment:
💭 Should probably get rid of this user interaction and do it in the UI instead. Not necessarily in this PR though.